x-pack/plugins/security_solution/public/resolver/store/data/selectors.ts

```diff
@@ -529,3 +557,90 @@ export const processNodePositionsAndEdgeLineSegments = createSelector(
     };
   }
 );
+
+const indexedProcessNodePositionsAndEdgeLineSegments = createSelector(
+  processNodePositionsAndEdgeLineSegments,
+  function visibleProcessNodePositionsAndEdgeLineSegments({
+    /* eslint-disable no-shadow */
+    processNodePositions,
+    edgeLineSegments,
+    /* eslint-enable no-shadow */
+  }) {
+    const tree: rbush<IndexedEntity> = new rbush();
+    const processesToIndex: IndexedProcessNode[] = [];
+    const edgeLineSegmentsToIndex: IndexedEdgeLineSegment[] = [];
+
+    const processNodeViewWidth = 720;
+    const processNodeViewHeight = 240;
+    const lineSegmentPadding = 30;
+    for (const [processEvent, position] of processNodePositions) {
+      const [nodeX, nodeY] = position;
+      const indexedEvent: IndexedProcessNode = {
+        minX: nodeX - 0.5 * processNodeViewWidth,
+        minY: nodeY - 0.5 * processNodeViewHeight,
+        maxX: nodeX + 0.5 * processNodeViewWidth,
+        maxY: nodeY + 0.5 * processNodeViewHeight,
+        position,
+        entity: processEvent,
+        type: 'processNode',
+      };
+      processesToIndex.push(indexedEvent);
+    }
+    for (const edgeLineSegment of edgeLineSegments) {
+      const {
+        points: [[x1, y1], [x2, y2]],
+      } = edgeLineSegment;
+      const indexedLineSegment: IndexedEdgeLineSegment = {
+        minX: Math.min(x1, x2) - lineSegmentPadding,
+        minY: Math.min(y1, y2) - lineSegmentPadding,
+        maxX: Math.max(x1, x2) + lineSegmentPadding,
+        maxY: Math.max(y1, y2) + lineSegmentPadding,
+        entity: edgeLineSegment,
+        type: 'edgeLine',
+      };
+      edgeLineSegmentsToIndex.push(indexedLineSegment);
+    }
+    tree.load([...processesToIndex, ...edgeLineSegmentsToIndex]);
+    return tree;
+  }
+);
+
+export const visibleProcessNodePositionsAndEdgeLineSegments = createSelector(
+  indexedProcessNodePositionsAndEdgeLineSegments,
+  function visibleProcessNodePositionsAndEdgeLineSegments(tree) {
+    // memoize the results of this call to avoid unnecessarily rerunning
+    let lastBoundingBox: AABB | null = null;
+    let currentlyVisible: VisibleEntites = {
+      processNodePositions: [],
+      connectingEdgeLineSegments: [],
+    };
+    return (boundingBox: AABB) => {
+      if (lastBoundingBox !== null && isEqual(lastBoundingBox, boundingBox)) {
+        return currentlyVisible;
+      } else {
+        const {
+          minimum: [minX, minY],
+          maximum: [maxX, maxY],
+        } = boundingBox;
+        const entities = tree.search({
+          minX,
+          minY,
+          maxX,
+          maxY,
+        });
+        const visibleProcessNodePositions = entities.filter(
```

Comment:
TODO, map these back to the non-indexed versions. This encapsulates the internal indexed structure. This is good because we don't actually need that structure, rbush can be configured to work w/ our `AABB` type.